ext/ldap/ldap.c

```diff
@@ -2491,351 +2491,255 @@ PHP_FUNCTION(ldap_delete_ext)
 }
 /* }}} */
 
-/* {{{ _ldap_str_equal_to_const */
-static size_t _ldap_str_equal_to_const(const char *str, size_t str_len, const char *cstr)
-{
-	size_t i;
-
-	if (strlen(cstr) != str_len)
-		return 0;
-
-	for (i = 0; i < str_len; ++i) {
-		if (str[i] != cstr[i]) {
-			return 0;
-		}
-	}
-
-	return 1;
-}
-/* }}} */
-
-/* {{{ _ldap_strlen_max */
-static size_t _ldap_strlen_max(const char *str, size_t max_len)
-{
-	size_t i;
-
-	for (i = 0; i < max_len; ++i) {
-		if (str[i] == '\0') {
-			return i;
-		}
-	}
-
-	return max_len;
-}
-/* }}} */
-
-/* {{{ _ldap_hash_fetch */
-static void _ldap_hash_fetch(zval *hashTbl, const char *key, zval **out)
-{
-	*out = zend_hash_str_find(Z_ARRVAL_P(hashTbl), key, strlen(key));
-}
-/* }}} */
-
 /* {{{ Perform multiple modifications as part of one operation */
 PHP_FUNCTION(ldap_modify_batch)
 {
-	zval *serverctrls = NULL;
+	zval *server_controls_zv = NULL;
 	ldap_linkdata *ld;
-	zval *link, *mods, *mod, *modinfo;
-	zend_string *modval;
-	zval *attrib, *modtype, *vals;
-	zval *fetched;
+	zval *link;
 	char *dn;
 	size_t dn_len;
-	int i, j, k;
-	int num_mods, num_modprops, num_modvals;
+	HashTable *modifications;
 	LDAPMod **ldap_mods;
 	LDAPControl **lserverctrls = NULL;
-	uint32_t oper;
 
 	/*
-	$mods = array(
-		array(
+	$mods = [
+		[
 			"attrib" => "unicodePwd",
 			"modtype" => LDAP_MODIFY_BATCH_REMOVE,
-			"values" => array($oldpw)
-		),
-		array(
+			"values" => [$old_pwd]
+		],
+		[
 			"attrib" => "unicodePwd",
 			"modtype" => LDAP_MODIFY_BATCH_ADD,
-			"values" => array($newpw)
-		),
-		array(
+			"values" => [$new_pwd]
+		],
+		[
 			"attrib" => "userPrincipalName",
 			"modtype" => LDAP_MODIFY_BATCH_REPLACE,
-			"values" => array("janitor@corp.contoso.com")
-		),
-		array(
+			"values" => ["janitor@corp.contoso.com"]
+		],
+		[
 			"attrib" => "userCert",
 			"modtype" => LDAP_MODIFY_BATCH_REMOVE_ALL
-		)
-	);
+		],
+	];
 	*/
 
-	if (zend_parse_parameters(ZEND_NUM_ARGS(), "Osa/|a!", &link, ldap_link_ce, &dn, &dn_len, &mods, &serverctrls) != SUCCESS) {
+	if (zend_parse_parameters(ZEND_NUM_ARGS(), "Osh/|a!", &link, ldap_link_ce, &dn, &dn_len, &modifications, &server_controls_zv) != SUCCESS) {
 		RETURN_THROWS();
 	}
 
 	ld = Z_LDAP_LINK_P(link);
 	VERIFY_LDAP_LINK_CONNECTED(ld);
 
 	/* perform validation */
-	{
-		zend_string *modkey;
-		zend_long modtype;
+	/* make sure the DN contains no NUL bytes */
+	if (zend_char_has_nul_byte(dn, dn_len)) {
+		zend_argument_value_error(2, "must not contain null bytes");
+		RETURN_THROWS();
+	}
```

Comment:
I will do a follow-up to use the `p` ZPP modifier consistently for the DN parameter.